Scripts/Runtime/Entities/Lifecycle/AbstractEntityLifecycleStatusSystem.cs

```diff
@@ -0,0 +1,303 @@
+            private readonly AccessController m_AccessController;
+
+            private NativeList<int> m_WorldEntitiesState;
+            private NativeList<Entity> m_WorldCreatedEntities;
+            private NativeList<Entity> m_WorldDestroyedEntities;
+            private NativeParallelHashSet<Entity> m_WorldCleanupEntities;
+            private NativeReference<int> m_CleanupFrame;
+
+            private EntityQuery m_CleanupEntityQuery;
+
+            private WorldEntityState(World world)
+            {
+                m_World = world;
+                m_LifecycleStatusSystems = new HashSet<AbstractEntityLifecycleStatusSystem>();
+                m_AccessController = new AccessController();
+
+                m_WorldEntitiesState = new NativeList<int>(ChunkUtil.MaxElementsPerChunk<Entity>(), Allocator.Persistent);
+                m_WorldCreatedEntities = new NativeList<Entity>(ChunkUtil.MaxElementsPerChunk<Entity>(), Allocator.Persistent);
+                m_WorldDestroyedEntities = new NativeList<Entity>(ChunkUtil.MaxElementsPerChunk<Entity>(), Allocator.Persistent);
+                m_WorldCleanupEntities = new NativeParallelHashSet<Entity>(ChunkUtil.MaxElementsPerChunk<Entity>(), Allocator.Persistent);
+                m_CleanupFrame = new NativeReference<int>(-1, Allocator.Persistent);
+
+                m_CleanupEntityQuery = m_World.EntityManager.CreateEntityQuery(EntityCleanupHelper.CLEAN_UP_ENTITY_COMPONENT_TYPE);
+            }
+
+            private void Dispose()
+            {
+                m_AccessController.Dispose();
+                m_WorldEntitiesState.Dispose();
+                m_WorldCreatedEntities.Dispose();
+                m_WorldDestroyedEntities.Dispose();
+                m_WorldCleanupEntities.Dispose();
+                m_CleanupFrame.Dispose();
+                m_CleanupEntityQuery.Dispose();
+            }
+
+            private void AddLifecycleStatusSystem(AbstractEntityLifecycleStatusSystem lifecycleStatusSystem)
+            {
+                m_LifecycleStatusSystems.Add(lifecycleStatusSystem);
+            }
+
+            public void RemoveLifecycleStatusSystem(AbstractEntityLifecycleStatusSystem lifecycleStatusSystem)
+            {
+                m_LifecycleStatusSystems.Remove(lifecycleStatusSystem);
+                if (m_LifecycleStatusSystems.Count == 0)
+                {
+                    Dispose();
+                }
+            }
+
+            public JobHandle AcquireCreatedAndDestroyedEntities(
+                JobHandle dependsOn,
+                out NativeArray<Entity>.ReadOnly createdEntities,
+                out NativeArray<Entity>.ReadOnly destroyedEntities,
+                bool shouldIncludeCleanupEntities)
+            {
+                m_AccessController.Acquire(AccessType.ExclusiveWrite);
+                
+                m_World.EntityManager.GetCreatedAndDestroyedEntities(m_WorldEntitiesState, m_WorldCreatedEntities, m_WorldDestroyedEntities);
```

Comment:
I don't know enough about this call but does this get the created/destroyed entities at the current moment or the ones that were created/destroyed at the start of the frame?

If it's current then having the Acquire method also fetch updated data is counter-intuitive.

Reply:
The reason for this is that Unity will operate on the underlying NativeLists on the main thread at the time of the call.

Imagine the scenario where you have two of these systems in the world at different points in the frame. One for Simulation, one for Rendering for example.

Sim goes first, the `GetCreatedAndDestroyedEntities` call happens and you get back the data arrays synchronously. 

You then schedule some jobs to operate on that data. 

If the Rendering one gets to go while those jobs are still in flight, those data arrays that they were operating on have now been modified. 

By using this acquire/release pattern, if any jobs were in flight, they'll get completed before we call `GetCreatedAndDestroyedEntities` again.